src/sm.c

```diff
@@ -0,0 +1,331 @@
+    if ((b_now & 1) == 0)
+    {
+        InterlockedExchange(&sm->e_done, 0);
+
+        LONG64 n = InterlockedIncrement64(&sm->n);
+        (void)n;
+
+        /*Codes_SRS_SM_02_030: [ sm_barrier_begin shall succeed and return SM_EXEC_GRANTED. ]*/
+        /*Codes_SRS_SM_02_017: [ sm_close_begin shall succeed and return SM_EXEC_GRANTED. ]*/
+        result = SM_EXEC_GRANTED;
+
+        /*drain previous executions*/
+        /*Codes_SRS_SM_02_016: [ sm_close_begin shall wait for all previous operations to end. ]*/
+        /*Codes_SRS_SM_02_029: [ sm_barrier_begin shall wait for the completion of all the previous operations. ]*/
+        while (get_n_minus_e(sm) != 1)
+        {
+            if (InterlockedHL_WaitForValue(&sm->e_done, 1, INFINITE) != INTERLOCKED_HL_OK)
+            {
+                /*Codes_SRS_SM_02_031: [ If there are any failures then sm_barrier_begin shall fail and return SM_ERROR. ]*/
+                /*Codes_SRS_SM_02_034: [ If there are any failures then sm_close_begin shall fail and return SM_ERROR. ]*/
+                LogError("Catastrophic failure: InterlockedHL_WaitForValue(&sm->e_done, 1, INFINITE), name=%s, n=%" PRId64 "", sm->name, n);
+                InterlockedIncrement64(&sm->e); /*this is pretty fatal here - the wait failed... */
+                result = SM_ERROR;
+                break;
+            }
+            else
+            {
+                /*some thread thinks that n - 1 == e, the while condition will look at that throughly*/
+            }
+        }
+    }
+    else
+    {
+        /*Codes_SRS_SM_02_015: [ If setting the lowest bit b_now to 1 fails then sm_close_begin shall return SM_EXEC_REFUSED. ]*/
+        result = SM_EXEC_REFUSED;
+    }
+    return result;
+}
+
+
+SM_RESULT sm_close_begin(SM_HANDLE sm)
+{
+
+    SM_RESULT result;
+    /*Codes_SRS_SM_02_013: [ If sm is NULL then sm_close_begin shall fail and return SM_ERROR. ]*/
+    if (sm == NULL)
+    {
+        LogError("invalid argument SM_HANDLE sm=%p", sm);
+        result = SM_ERROR;
+    }
+    else
+    {
+        result = sm_barrier_begin_internal(sm);
+    }
+    return result;
+}
+
+void sm_close_end(SM_HANDLE sm)
+{
+    /*Codes_SRS_SM_02_018: [ If sm is NULL then sm_close_end shall return. ]*/
+    if (sm == NULL)
+    {
+        LogError("invalid argument SM_HANDLE sm=%p", sm);
+    }
+    else
+    {
+        /*cannot really check that it is paired with a _close_begin without moving b_now through user land, not going to do that. That solution woould push the b_now of the "close" in user land. And user has to return with the same b_now here and check that it is the same*/
+        /*Codes_SRS_SM_02_019: [ sm_close_end shall switch b_now to -1, n to 0 and e to 0. ]*/
+        (void)InterlockedExchange64(&sm->b_now, -1);
+        (void)InterlockedExchange64(&sm->n, 0);
+        (void)InterlockedExchange64(&sm->e, 0);
+        (void)InterlockedExchange(&sm->e_done, 0);
+    }
+}
+
+SM_RESULT sm_begin(SM_HANDLE sm)
```

Comment:
Probably not actionable in this review:

When I wrote my first attempt I went down the same route: having the user of the state manager call into a pair of functions to mark where the operation (like a write) starts and end.
That obviously still requires the user to write something like:

if (sm_begin() != OK)
{
   ... log error ...
}
else
{
    .. do some work ...
    sm_end() <- this gets executed if a later error happened, otherwise there is a similar call in the callback
}

I tried to think of a better way to tie together the begin/end so that they do not accidentally get mismatched but I could not really come up with anything interesting. Maybe you have better ideas.

Reply:
Sort of RAII like maybe with some macros.

---
In reply to: [421816537](https://github.com/Azure/azure-c-util/pull/35#discussion_r421816537) [](ancestors = 421816537)